keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -0,0 +1,1001 @@
+
+1. Enable 3-tier policy admission plugin, default everything to privileged.
+2. Eliminate mutating PSPs:
+    1. Clone all mutating PSPs to a non-mutating version (automated)
+    2. Update all ClusterRoles authorizing use of the mutating PSPs to also authorize use of the
+       non-mutating variant (automated)
+    3. Watch for pods using the mutating PSPs (check via the `kubernetes.io/psp` annotation), and
+       work with code owners to migrate to valid non-mutating resources.
+    4. Delete mutating PSPs
+3. Select a compatible profile for each namespace, based on the existing resources in the namespace
+   (automated)
+    1. Review the profile choices
+    2. Evaluate the difference in privileges that would come from disabling the PSP controller
+       (automated).
+4. (optional) Apply the profiles in `warn` and `audit` mode (automated)
+5. Apply the profiles in `allow` mode (automated)
+6. Disable PodSecurityPolicy
+
+### Custom Profiles
+
+Allow custom profile levels to be statically configured. E.g.
+`--extra-pod-security-levels=host-network`. Custom profiles are ignored by the built-in admission
+plugin, and must be handled completely by a 3rd party webhook (including the dry-run implementation,
+if desired).
+
+### Custom Warning Messages
+
+An optional `prefix/warn-message` annotation can be used to return a custom warning message (in
+addition to the standard message) whenever a policy warning is triggered. This could be useful for
+announcing the date that new policy will take effect, or providing a point of contact if you need to
+request an exception.
+
+### Windows Support
+
+Even without built-in support, enforcement for Windows pods can be delegated to a webhook admission
+plugin by exempting the `windows` RuntimeClass. We should investigate built-in Windows support out
+of the box though. Requirements include:
+
+1. A standardized identifier for Windows pods
+2. Write the Pod Security Standards for windows
+
+Risk: If a Windows RuntimeClass uses a runtime handler that is also configured on linux nodes, then
+a user can just create a linux pod with the Windows RuntimeClass and manually schedule it to a linux
+node to bypass the policy checks. For example, this would be the case if the cluster was exclusively
+using the dockershim runtime, which requires the hardcoded `docker` runtime handler to be set.
+
+### Offline Policy Checking
+
+We could provide a standalone tool that is capable of checking the policies against resource files
+or through stdin. It should be capable of evaluating `AdmissionReview` resources, but also pod and
+templated pod resources. This could be useful in CI/CD pipelines and tests.
+
+### Event recording
+
+Allow recording an event in response to a pod creation attempt that exceeds a given level.
+
+### Conformance
+
+As this feature progresses towards GA, we should think more about how it interacts with conformance.
```

Comment:
Maybe also think about how to provide conformance tests that allow rival implementations?

For example (just a sketch): various Pods and templates that either violate or comply with different Pod Security Standards, and expected admission control results for a policy enforcing said standard.

Reply:
I agree, but this is sort of covered under the testing section for "Manual Testing Resources". I don't want to call that conformance right now, since conformance implies a formality that I'm not ready to take on with these policies yet.

Reply:
What that needs to look like and if it needs to allow significantly different implementations is a question for the conformance definition working group and can be determined later. I personally think there's value in having a standard here but we have a separate venue for this. cc @johnbelamaric @spiffxp

Reply:
Conformance tests kube apis and the behavior of them.  Having an API that others are expected to follow should never test the implementation (we are very careful to for instance spec how a container runtime should behave AS a pod would see it, but not tie it to a container runtime).

You may not require an implementation of an API in general in conformance, just define the behavior that you expect to be required for conformance (and we bias towards caution on requiring that).